lib/streamlit/connections/snowflake_connection.py

```diff
@@ -0,0 +1,252 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# NOTE: We won't always be able to import from snowflake.{connector, snowpark}.* so need
+# the `type: ignore` comment below, but that comment will explode if `warn-unused-ignores`
+# is turned on when the package is available. Unfortunately, mypy doesn't provide a good
+# way to configure this at a per-line level :(
+# mypy: no-warn-unused-ignores
+
+from __future__ import annotations
+
+from datetime import timedelta
+from typing import TYPE_CHECKING, Any, cast
+
+import pandas as pd
+
+from streamlit.connections import BaseConnection
+from streamlit.connections.util import running_in_sis
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.caching import cache_data
+
+if TYPE_CHECKING:
+    from snowflake.connector import (  # type:ignore[import] # isort: skip
+        SnowflakeConnection as InternalSnowflakeConnection,
+    )
+    from snowflake.connector.cursor import SnowflakeCursor  # type:ignore[import]
+    from snowflake.snowpark.session import Session  # type:ignore[import]
+
+
+_REQUIRED_CONNECTION_PARAMS = {"account"}
+
+
+def _validate_connection_params(required: set[str], params: dict[str, Any]) -> None:
+    missing = required - set(params)
+    if len(missing) > 0:
+        raise StreamlitAPIException(
+            f"Missing Snowflake connection params: {', '.join(missing)}"
+        )
+
+
+class SnowflakeConnection(BaseConnection["InternalSnowflakeConnection"]):
+    """A connection to Snowflake using the Snowflake Python Connector. Initialize using
+    ``st.connection("<name>", type="snowflake")``.
+
+    SnowflakeConnection supports direct SQL querying using ``.query("...")``, access to
+    the underlying Snowflake Python Connector object with ``.raw_connection``, and other
+    convenience functions. See the methods below for more information.
+    SnowflakeConnections should always be created using ``st.connection()``, **not**
+    initialized directly.
+    """
+
+    def _connect(self, **kwargs) -> "InternalSnowflakeConnection":
+        import snowflake.connector  # type:ignore[import]
+        from snowflake.snowpark.context import get_active_session  # type:ignore[import]
+
+        # If we're running in SiS, just call get_active_session() and retrieve the
+        # lower-level connection from it.
+        if running_in_sis():
+            session = get_active_session()
+
+            if hasattr(session, "connection"):
+                return session.connection
```

Comment:
Will do -- I'll go ahead and click merge on this PR to avoid having to wait on another CI run, but I'll add these comments in a followup PR for docstring tweaks and comment adjustments